rocAL/include/readers/image/numpy_data_reader.h

```diff
@@ -0,0 +1,134 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#pragma once
+#include <dirent.h>
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "pipeline/commons.h"
+#include "pipeline/timing_debug.h"
+#include "readers/image/image_reader.h"
+
+class NumpyDataReader : public Reader {
+   public:
+    //! Looks up the folder which contains the files, amd loads the numpy files
+    /*!
+     \param desc  User provided descriptor containing the files' path.
+    */
+    Reader::Status initialize(ReaderConfig desc) override;
+    //! Reads the next resource item
+    /*!
+     \param buf User's provided buffer to receive the loaded numpy data
+     \return Size of the loaded resource
+    */
+    size_t read_data(unsigned char* buf, size_t max_size) override;
+    //! Opens the next file in the folder
+    /*!
+     \return The size of the next file, 0 if couldn't access it
+    */
+    size_t open() override;
+
+    //! Returns the numpy header data 
+    /*!
+     \return The numpy header data of the current file
+    */
+    const NumpyHeaderData get_numpy_header_data() override;
+
+    //! Reads the next numpy file
+    /*!
+     \param buf User's provided buffer to receive the loaded numpy data
+     \param raad_size Requested read size by the user
+     \param buf Max dimension strides of the output tensor used for padding along the dimensions
+     \return Size of the loaded resource
+    */
+    size_t read_numpy_data(void* buf, size_t read_size, std::vector<unsigned>& max_shape) override;
+
+    //! Resets the object's state to read from the first file in the folder
+    void reset() override;
+
+    //! Returns the name of the latest file opened
+    std::string id() override { return _last_id; };
+
+    //! Returns the name of the latest file_path opened
+    const std::string file_path() override { return _last_file_path; }
+
+    unsigned count_items() override;
+
+    ~NumpyDataReader() override;
+
+    int close() override;
+
+    NumpyDataReader();
+
+    std::string get_root_folder_path() override;  // Returns the root folder path
+
+   private:
+    //! opens the folder containing the numpy arrays
+    Reader::Status open_folder();
+    Reader::Status subfolder_reading();
+    std::string _folder_path;
+    std::string _file_list_path;
+    DIR *_src_dir;
+    DIR *_sub_dir;
+    struct dirent *_entity;
+    std::vector<std::string> _file_names;
+    std::vector<std::string> _files;
+    std::vector<NumpyHeaderData> _file_headers;
+    FILE *_current_fPtr;
+    unsigned _current_file_size;
+    NumpyHeaderData _curr_file_header;
+    std::string _last_id;
+    std::string _last_file_name, _last_file_path, _absolute_file_path;
+    size_t _batch_size = 1;
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    unsigned _seed = 0;
```

Comment:
loop, shuffle, seed, 
All are present as part of the Reader class
These changes are part of the C++ test suite fix PR
We might need to merge that before this PR, so we can have the Reader changes from that PR

Reply:
I will merge the changes once its pulled into ToT